Feature/update11.0

eironside opened the pull request:
Updates for ArcGIS GeoEvent Server 11.0 and fix for issues related to investigating [#3319 Refactor the Track Idle Detector on the GeoEvent Server Gallery](https://devtopia.esri.com/WebGIS/arcgis-geoevent/issues/3319)

**NOTE: The main request of 3319 was that Track Idle Detector include Track Gap functionality. That specific functional request has been identified as an enhancement and will be evaluated under a separate effort.**

#### **Updates for 11.0**
- Formatted all files to current file format standards. Fixed line indents, spacing and other general formatting issues.
- Standard updates to POM files, parent/child, updated dependency versions
- Improved processor description and property descriptions

#### **Functional updates**
- Duplicate field names in incoming event definition [idle, idleDuration, idleStart] are now logged and event is dropped
- Events with definition that doesn't have GEOMETRY, TIME_START, or TRACK_ID tags set are dropped and logged as error
- Events with null field values for GEOMETRY, TIME_START, or TRACK_ID fields are dropped and logged as error
- Fixed issue with collision for TRACK_ID and GEOMETRY fields when keep fields is enabled (if TRACK_ID and GEOMETRY exist in incoming event definition, those fields are updated rather than adding new fields).
 - Updated logging strategy
      - Added trace logging to allow user to follow event processing logic
      - Added more descriptive log error messages so user knows how to fix issues
      - Updated error logging to gate stack trace for DEBUG only but still use WARN level
      - Gate any log message that requires method evaluation (e.g. _LOG.trace("def name {0}", definition.getName())_ )

eironside committed:
File format changes (no functional changes)

eironside committed:
Standard Updates To POM/README/Strings/Definition

Update POM Structures
Added revision property to manage SDK version
Added project.release property
Updated dependency versions to latest
Migrated standard functionality to parent POM
Added detail to string descriptions and fixed errors
Added project.release to processor description
Updated README.md to include build instructions, requirements, resources

eironside committed:
fixed line indent

eironside committed:
Update Logging & fix field collision issue



Fixed URL to gallery item in parent POM file
Updating exception logging strategy
Fixed copy/paste error in log message 'Gap Time' to 'Idle Limit'
Added trace logging for troubleshooting
Added cacheKey to distance calc to allow log to report TRACK_ID that is
idle
Fixed issue with adding incoming schema fields having TRACK_ID and
GEOMETRY on different fields
Removed TrackIdleDetectorDefinition logger
Added toString() on TrackIdleState
Fixed and enhanced label, description, and logging strings

eironside committed:
Added duplicate feild error and null/missing field error

TrackIdleDetector.java
 - Updated exception log strategy (again)
 - Added check for null/missing TIME_START per requirements
 - Improved log message for missing fields (need a tag or the field is
null)
 - Keep fields: Moved definition creation to just in time
 - Added check for duplicate fields (drop event and throw 'duplicate
field name' exception)

TrackIdleDetectorDefinition.java
 - Added getName() and getDomain() back to make logging work

trackidledetector-processor.properties
 - Added escape char for quotes to make logs work
 - Updated null field/missing tag message to be clear for user
 - Added name of definition that failed to be added to Definition
Manager
 - Added duplicate field error message

hanoch commented on pom.xml:
```diff
@@ -1,66 +1,95 @@
-	<version>10.6.0</version>
-	<packaging>pom</packaging>
+  <groupId>com.esri.geoevent.parent</groupId>
+  <artifactId>trackidledetector</artifactId>
```
We should not change the artifact id

eironside replied:
Needs to match the path to the resources string folder/file. Otherwise strings don't load nor get replaced by the resources bundle.

hanoch commented on pom.xml:
```diff
@@ -1,66 +1,95 @@
 
-	<name>Esri :: GeoEvent :: TrackIdleDector</name>
-	<url>http://www.esri.com</url>
+  <name>${project.artifactId}-parent</name>
```
We should not change the bundle name format convention. All GES OSGi bundle names are formatted with the `A :: B` convention.

eironside replied:
Using the ${artifactId} makes it easier to manage if the name ever does change.  But that doesn't happen often, so I can change it back.

hanoch commented on trackidledetector-processor/pom.xml:
```diff
@@ -1,41 +1,43 @@
-			<plugin>
-				<groupId>org.apache.maven.plugins</groupId>
-				<artifactId>maven-compiler-plugin</artifactId>
-				<configuration>
-					<source>1.8</source>
-					<target>1.8</target>
-				</configuration>
```
Was that section removed or moved?

eironside replied:
Moved to parent.  Maven pom.xml best practices say to put as much as possible in the root/parent pom file under 'Management' sections then only use as necessary in the child pom(s).

hanoch commented on trackidledetector-processor/src/main/java/com/esri/geoevent/processor/trackidledetector/TrackIdleDetectorDefinition.java:
```diff
@@ -24,84 +24,86 @@
+  @Override
+  public String getName()
+  {
+    return "trackidledetector-processor";
```
please keep the CamelCase (hungarian) notation

eironside replied:
needs to match the resources path to the strings file otherwise strings aren't resolved by the logging system at runtime (e.g. "${com.esri.geoevent.processor.trackidledetector-processor.PROCESSOR_DESC}")

hanoch commented on trackidledetector-processor/src/main/resources/com/esri/geoevent/processor/trackidledetector-processor.properties:
```diff
@@ -1,26 +1,46 @@
 # Processor Definition
 PROCESSOR_LABEL=Track Idle Detector
-PROCESSOR_DESC=Detects non-movement of a Track beyond a specified period of time.
+PROCESSOR_DESC=Release ${project.release}: Detects non-movement of a Track beyond a specified period of time.
```
Release???

eironside replied:
Version is the GeoEvent SDK version that the bundle was compiled against. And we don't change that if a bundle is enhanced or has a bug fixed, so we need another way to indicate a bundle's version. So I added the project.release property to indicate a bundle's release (aka version) that is independant of the SDK version.

eironside replied:
See the following for example of how the release is used:

[Track Idle Detector Processor Gallery Item page](https://geoevent.maps.arcgis.com/home/item.html?id=5d8e3446736d4df299c7c96bc275d561)

eironside committed:
Fixed POM Names

Update main pom.xml name to Esri :: GeoEvent :: TrackIdleDector
Fixed .jar file final name (name->artifactid)
Update child pom.xml name to Esri :: GeoEvent :: TrackIdleDector
Processor
